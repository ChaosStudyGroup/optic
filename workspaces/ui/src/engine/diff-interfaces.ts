interface SerializedDiff {
  [key: string]: {
    interactionTrail: {
      path: any[];
    };
    requestsTrail: any;
  };
}

// Diff Types the UI Handles

export const allowedDiffTypes: {
  [key: string]: { isBodyShapeDiff: boolean };
} = {
  UnmatchedRequestUrl: { isBodyShapeDiff: false },
  UnmatchedRequestMethod: {
    isBodyShapeDiff: false,
  },
  UnmatchedRequestBodyContentType: {
    isBodyShapeDiff: false,
  },
  UnmatchedResponseBodyContentType: {
    isBodyShapeDiff: false,
  },
  UnmatchedResponseStatusCode: {
    isBodyShapeDiff: false,
  },
  UnmatchedRequestBodyShape: {
    isBodyShapeDiff: true,
  },
  UnmatchedResponseBodyShape: {
    isBodyShapeDiff: true,
  },
};

export const allowedDiffTypesKeys: string[] = Object.keys(allowedDiffTypes);

// Properties of Each Diff Types

export const isBodyShapeDiff = (key: string): boolean =>
  allowedDiffTypes[key]?.isBodyShapeDiff;
export const DiffInRequest = (key: string): boolean =>
  key === 'UnmatchedRequestBodyShape';
export const DiffInResponse = (key: string): boolean =>
  key === 'UnmatchedResponseBodyShape';

// The ones we like to work with in the UI

export interface BodyShapeDiff<ShapeDiffLocation> {
  diff: any;
  type: string;
  interactionPointers: string[];
  location: Location;
}

export interface RequestBodyLocation {
  contentType: string;
}

export interface ResponseBodyLocation {
  statusCode: number;
  contentType: string;
}
