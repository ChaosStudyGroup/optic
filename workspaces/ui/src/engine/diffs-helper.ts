import invariant from 'invariant';
import {
  allowedDiffTypesKeys,
  BodyShapeDiff,
  DiffInRequest,
  DiffInResponse,
  isBodyShapeDiff,
  RequestBodyLocation,
} from './interfaces/interfaces';
import {
  IUnmatchedRequestBodyShape,
  IUnmatchedResponseBodyShape,
} from './interfaces/diffs';

class DiffLifecycleManager {
  constructor(private serialized_diff: any, interactions: string[]) {}

  diffType(): string {
    const keys = Object.keys(this.serialized_diff);
    const typeKey = keys[0];
    invariant(
      keys.length === 1 && allowedDiffTypesKeys.includes(typeKey),
      'Serialized diffs should only have one root key'
    );
    return typeKey!;
  }

  isShapeDiff(): boolean {
    return isBodyShapeDiff(this.diffType());
  }

  inRequest(): boolean {
    return DiffInRequest(this.diffType());
  }

  inResponse(): boolean {
    return DiffInResponse(this.diffType());
  }

  asRequestBodyShapeDiff: BodyShapeDiff<
    IUnmatchedRequestBodyShape,
    RequestBodyLocation
  > = {};

  asResponseBodyShapeDiff: BodyShapeDiff<
    IUnmatchedResponseBodyShape,
    RequestBodyLocation
  > = {};
}
