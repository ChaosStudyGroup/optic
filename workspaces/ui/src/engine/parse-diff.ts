import {
  IDiff,
  IDiffWithShapeDiff,
  IShapeDiffResult,
  IUnmatchedRequestBodyShape,
  IUnmatchedResponseBodyShape,
} from './interfaces/diffs';
import jsonStringify from 'json-stable-stringify';
import sha1 from 'node-sha1';
import {
  allowedDiffTypes,
  allowedDiffTypesKeys,
  DiffInRequest,
  IParsedLocation,
  IRequestBodyLocation,
  IResponseBodyLocation,
  isBodyShapeDiff,
} from './interfaces/interfaces';
import invariant from 'invariant';
import { IShapeTrail } from './interfaces/shape-trail';

export class ParsedDiff {
  diffType: string;

  constructor(private serialized_diff: IDiff, private interactions: string[]) {
    const keys = Object.keys(this.serialized_diff);
    const typeKey = keys[0]!;
    invariant(
      keys.length === 1 && allowedDiffTypesKeys.includes(typeKey),
      'Serialized diffs should only have one root key'
    );

    this.diffType = typeKey!;
  }

  raw(): IDiff {
    return this.serialized_diff;
  }

  location(): IParsedLocation {
    return {
      pathId: 'xyz',
      method: 'GET',
      inRequest: DiffInRequest(this.diffType) && null,
      inResponse: DiffInRequest(this.diffType) && null,
    };
  }

  asShapeDiff(): BodyShapeDiff | undefined {
    if (!isBodyShapeDiff(this.diffType)) {
      return undefined;
    }

    const asWithShapeDiff = this.serialized_diff as IDiffWithShapeDiff;

    const requestBodyShapeDiff: IShapeDiffResult | undefined =
      asWithShapeDiff[allowedDiffTypes.UnmatchedRequestBodyShape.asString]
        ?.shapeDiffResult;

    const responseBodyShapeDiff: IShapeDiffResult | undefined =
      asWithShapeDiff[allowedDiffTypes.UnmatchedResponseBodyShape.asString]
        .shapeDiffResult;

    return new BodyShapeDiff(
      this,
      this.serialized_diff,
      (requestBodyShapeDiff || responseBodyShapeDiff)!,
      this.interactions,
      this.location()
    );
  }
}

export function parseDiffsArray(array: [IDiff, string[]][]): ParsedDiff[] {
  return array.map(([rawDiff, pointers]) => new ParsedDiff(rawDiff, pointers));
}

class BodyShapeDiff {
  shapeTrail: IShapeTrail;
  shapeTrailHash: string;

  constructor(
    private parsedDiff: ParsedDiff,
    private diff: IDiff,
    private shapeDiff: IShapeDiffResult,
    public interactionPointers: string[],
    public location: IParsedLocation
  ) {
    this.shapeTrail = (shapeDiff['UnmatchedShape']?.shapeTrail ||
      shapeDiff['UnspecifiedShape']?.shapeTrail)!;
    invariant(
      this.shapeTrail,
      'A shape trail must be specified with all shape diffs'
    );
    this.shapeTrailHash = sha1(jsonStringify(this.shapeTrail));
  }
}
