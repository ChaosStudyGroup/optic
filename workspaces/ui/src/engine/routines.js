import { RfcCommands } from '@useoptic/domain';

export function commandsForUpdatingContribution(id, key, value) {
  return [RfcCommands.AddContribution(id, key, value)];
}
