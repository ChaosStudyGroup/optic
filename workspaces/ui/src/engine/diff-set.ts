import { ParsedDiff } from './parse-diff';

export class DiffSet {
  constructor(private diffs: ParsedDiff[]) {}

  count(): number {
    return this.diffs.length;
  }

  forEndpoint(pathId: string, method: string): DiffSet {
    return new DiffSet(
      this.diffs.filter(
        (i) => i.location().pathId === pathId && i.location().method === method
      )
    );
  }
  newRegions(): DiffSet {
    return new DiffSet(this.diffs.filter((i) => !Boolean(i.asShapeDiff())));
  }
  shapeDiffs(): DiffSet {
    return new DiffSet(this.diffs.filter((i) => Boolean(i.asShapeDiff())));
  }

  iterator(): ParsedDiff[] {
    return this.diffs;
  }
}
