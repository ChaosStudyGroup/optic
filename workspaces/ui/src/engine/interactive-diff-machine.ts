import { BodyShapeDiff, ParsedDiff } from './parse-diff';
import { assign, Machine, send, sendParent } from 'xstate';
import { IShapeTrail } from './interfaces/shape-trail';
import {
  IDiffSuggestionPreview,
  initialTitleForNewRegions,
  prepareNewRegionDiffSuggestionPreview,
  prepareShapeDiffSuggestionPreview,
} from '../__tests__/diff-helpers/prepare-diff-previews';
import { InteractiveSessionConfig } from './interfaces/session';
import { IDiffDescription } from './interfaces/interpretors';
import { descriptionForDiffs } from './interpretors/DiffDescriptionInterpretors';
import {
  ILearnedBodies,
  IValueAffordanceSerializationWithCounter,
} from '@useoptic/cli-shared/build/diffs/initial-types';

interface DiffStateSchema {
  states: {
    unfocused: {};
    loading: {};
    ready: {
      states: {
        unhandled: {};
        handled: {};
      };
    };
  };
}

// The events that the machine handles
type DiffEvent =
  | { type: 'SHOWING' }
  | { type: 'CHANGE_SUGGESTION'; to: string }
  | { type: 'DISMISS_TYPE'; type_slug: string }
  | { type: 'STAGE' }
  | { type: 'UNSTAGE' }
  | { type: 'RESET' };

// The context (extended state) of the machine
interface DiffContext<InterpretationContext> {
  results: InterpretationContext | undefined;
  preview: IDiffSuggestionPreview | undefined;
  descriptionWhileLoading: IDiffDescription;
}

export const createNewRegionMachine = (
  id: string,
  parsedDiff: ParsedDiff,
  services: InteractiveSessionConfig
) =>
  Machine<DiffContext<ILearnedBodies>, DiffStateSchema, DiffEvent>({
    id,
    context: {
      results: undefined,
      preview: undefined,
      descriptionWhileLoading: descriptionForDiffs([parsedDiff]),
    },
    initial: 'unfocused',
    states: {
      unfocused: {
        on: {
          //@ts-ignore
          'done.invoke.loading-initial-regions': {
            actions: assign({
              //@ts-ignore
              results: (context, event) => event.data,
            }),
            target: 'loading',
          },
        },
      },
      loading: {
        invoke: {
          id: 'preparing-preview-of-diff',
          src: async (context, event) => {
            return prepareNewRegionDiffSuggestionPreview(
              parsedDiff,
              services,
              context.results!
            );
          },
          onDone: {
            target: 'ready',
            actions: assign({
              preview: (context, event) => event.data,
            }),
          },
        },
      },
      ready: {
        initial: 'unhandled',
        states: {
          unhandled: {
            entry: [
              (context, event) => console.log('ready to show the diff. '),
            ],
          },
          handled: {},
        },
      },
    },
  });
//
export const createShapeDiffMachine = (
  shapeTrail: IShapeTrail,
  diffs: ParsedDiff[],
  id: string,
  services: InteractiveSessionConfig
) =>
  Machine<
    DiffContext<IValueAffordanceSerializationWithCounter>,
    DiffStateSchema,
    DiffEvent
  >({
    id: id,
    context: {
      results: undefined,
      descriptionWhileLoading: descriptionForDiffs(diffs),
      preview: undefined,
    },
    initial: 'unfocused',
    states: {
      unfocused: {
        on: { SHOWING: 'loading' },
      },
      loading: {
        invoke: {
          id: 'loading-shape-diff-affordances',
          src: async (context, event) => {
            const firstDiff = diffs[0]!;
            const { pathId, method } = firstDiff.location();

            const trailValues = await services.diffService.learnTrailValues(
              services.rfcBaseState.rfcService,
              services.rfcBaseState.rfcId,
              pathId,
              method,
              firstDiff.raw()
            );

            return {
              preview: await prepareShapeDiffSuggestionPreview(
                diffs,
                services,
                trailValues
              ),
              results: trailValues,
            };
          },
          onDone: {
            target: 'ready',
            actions: assign({
              results: (context, event) => event.data.results, // cached in case we need to run 'em again
              preview: (context, event) => event.data.preview,
            }),
          },
        },
      },
      ready: {
        initial: 'unhandled',
        states: {
          unhandled: {
            entry: [(context, event) => console.log('ready to show the diff.')],
          },
          handled: {},
        },
      },
    },
  });
