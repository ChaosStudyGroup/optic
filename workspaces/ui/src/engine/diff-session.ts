import { assign, spawn, Machine, StateMachine, send } from 'xstate';
import { ParsedDiff } from './parse-diff';
import { DiffSet } from './diff-set';
import equals from 'lodash.isequal';
import { InteractiveSessionConfig } from './interfaces/session';
import {
  InteractiveEndpointSessionContext,
  InteractiveEndpointSessionEvent,
  newInteractiveEndpointSessionMachine,
} from './interactive-endpoint';
import { IUnrecognizedUrl } from '../services/diff';
import { IToDocument } from './interfaces/interfaces';

export interface DiffSessionSessionStateSchema {
  states: {
    loading: {};
    preparing: {};
    ready: {};
  };
}

// The events that the machine handles
export type DiffSessionSessionEvent =
  | { type: 'COMPLETED_DIFF'; diffs: ParsedDiff[]; urls: IUnrecognizedUrl[] }
  | { type: 'SHOWING' }
  | {
      type: 'SELECTED_ENDPOINT';
      pathId: string;
      method: string;
    }
  | {
      type: 'SEND_TO_ENDPOINT';
      pathId: string;
      method: string;
      event: InteractiveEndpointSessionEvent;
    }
  | { type: 'HANDLED_UPDATED'; pathId: string; method: string }
  | {
      type: 'TOGGLED_UNDOCUMENTED';
      active: boolean;
    }
  | {
      type: 'UPDATED_TO_DOCUMENT';
      handled: number;
      total: number;
      endpoints: { method: string; pathId: string; pathExpression: string }[];
      toDocument: IToDocument[];
    };

export function sendMessageToEndpoint(
  pathId: string,
  method: string,
  event: InteractiveEndpointSessionEvent
) {
  return {
    type: 'SEND_TO_ENDPOINT',
    pathId,
    method,
    event,
  };
}

export interface IEndpointWithStatus {
  pathId: string;
  method: string;
  handled: boolean;
  diffCount: number;
  machine: any;
  ref?: any;
}

// The context (extended state) of the machine
export interface DiffSessionSessionContext {
  allDiffs: ParsedDiff[];
  unrecognizedUrls: IUnrecognizedUrl[];
  endpoints: IEndpointWithStatus[];
  handledByEndpoint: IHandledByEndpoint;
  undocumentedEndpoints: { pathId: string; method: string }[];
  focus?: { pathId: string; method: string };
  showingUndocumented: boolean;
  unrecognizedUrlsToDocument: {
    urls: IToDocument[];
    endpoints: { pathId: string; method: string; pathExpression: string }[];
    handled: number;
    total: number;
  };
}

export const newDiffSessionSessionMachine = (
  diffId: string,
  services: InteractiveSessionConfig
) => {
  return Machine<
    DiffSessionSessionContext,
    DiffSessionSessionStateSchema,
    DiffSessionSessionEvent
  >({
    id: diffId,
    context: {
      handledByEndpoint: [],
      unrecognizedUrls: [],
      undocumentedEndpoints: [],
      allDiffs: [],
      endpoints: [],
      unrecognizedUrlsToDocument: {
        handled: 0,
        total: 0,
        endpoints: [],
        urls: [],
      },
      showingUndocumented: false,
    },
    initial: 'loading',
    states: {
      loading: {
        on: {
          COMPLETED_DIFF: {
            actions: assign({
              allDiffs: (context, event) => event.diffs,
              unrecognizedUrls: (context, event) => event.urls,
            }),
            target: 'preparing',
          },
        },
      },
      preparing: {
        invoke: {
          id: 'preparing-endpoints',
          src: async (context, event) => {
            const set = new DiffSet(context.allDiffs, services.rfcBaseState);

            const byEndpoint = set.groupedByEndpoint();
            const undocumented = set.forUndocumented();

            const undocumentedEndpoints = [];
            undocumented.iterator().forEach((i) => {
              const location = i.location(services.rfcBaseState);
              const entry = {
                method: location.method,
                pathId: location.pathId,
              };
              if (!undocumentedEndpoints.find((i) => equals(i, entry)))
                undocumentedEndpoints.push(entry);
            });

            const endpoints: IEndpointWithStatus[] = byEndpoint.map(
              ({ pathId, method, diffs }) => {
                return {
                  pathId,
                  handled: false,
                  diffCount: diffs.length,
                  method,
                  machine: newInteractiveEndpointSessionMachine(
                    pathId,
                    method,
                    diffs,
                    services
                  ),
                };
              }
            );
            return { endpoints, undocumentedEndpoints };
          },
          onDone: {
            target: 'ready',
            actions: [
              assign({
                undocumentedEndpoints: (_, event) =>
                  event.data.undocumentedEndpoints,
                endpoints: (context, event) => {
                  return event.data.endpoints.map((endpoint) => {
                    return {
                      ...endpoint,
                      ref: spawn(
                        endpoint.machine,
                        `${endpoint.pathId}${endpoint.method}`
                      ),
                    };
                  });
                },
              }),
            ],
          },
        },
      },
      ready: {
        on: {
          HANDLED_UPDATED: {
            actions: [
              assign({
                handledByEndpoint: (context) => computeHandled(context),
              }),
            ],
          },
          SEND_TO_ENDPOINT: {
            actions: (context, event) => {
              const actor = context.endpoints.find(
                (i) => i.pathId === event.pathId && i.method === event.method
              );
              if (actor) {
                actor.ref.send(event.event);
              }
            },
          },
          TOGGLED_UNDOCUMENTED: {
            actions: assign({
              showingUndocumented: (context, event) => event.active,
              focus: (ctx, event) => undefined,
            }),
          },
          UPDATED_TO_DOCUMENT: {
            actions: assign({
              unrecognizedUrlsToDocument: (context, event) => ({
                urls: event.toDocument,
                endpoints: event.endpoints,
                handled: event.handled,
                total: event.total,
              }),
            }),
          },
          SELECTED_ENDPOINT: {
            actions: [
              assign({
                focus: (context, event) => ({
                  pathId: event.pathId,
                  method: event.method,
                }),
                showingUndocumented: (ctx, event) => false,
              }),
              (context) => {
                const endpointMachine = context.endpoints.find(
                  (i) =>
                    i.pathId === context.focus.pathId &&
                    i.method === context.focus.method
                )!.ref;
                send({ type: 'PREPARE' }, { to: endpointMachine }); //warm it up
              },
            ],
          },
        },
      },
    },
  });
};

type IHandledByEndpoint = {
  pathId: string;
  method: string;
  diffCount: number;
  handled: number;
}[];
function computeHandled(
  context: DiffSessionSessionContext
): IHandledByEndpoint {
  const results: IHandledByEndpoint = [];

  context.endpoints.forEach((i) => {
    const endpointContext: InteractiveEndpointSessionContext =
      i.ref.state.context;
    const entries: [string, boolean][] = Object.entries(
      endpointContext.handledByDiffHash
    );
    const handled = entries.filter(([key, bool]) => bool).length;
    results.push({
      pathId: i.pathId,
      method: i.method,
      handled,
      diffCount: entries.length !== 0 ? entries.length : i.diffCount,
    });
  });

  return results;
}

export function nextEndpointToFocusOn(
  context: DiffSessionSessionContext,
  endpointsWithDiffs: { pathId: string; method: string }[]
) {
  const indexOfCurrentFoucs: number =
    context.focus &&
    endpointsWithDiffs.findIndex(
      ({ pathId, method }) =>
        pathId === context.focus.pathId && method === context.focus.method
    );

  function alreadyHandled(pathId: string, method: string) {
    const status = context.handledByEndpoint.find(
      (i) => pathId === i.pathId && method === i.method
    );
    return (
      status && status.diffCount > 0 && status.diffCount === status.handled
    );
  }
  // next unhandled after this one, or start from top of list.
  const searchOrder = [
    ...endpointsWithDiffs.slice(
      indexOfCurrentFoucs || 0,
      endpointsWithDiffs.length
    ),
    ...endpointsWithDiffs.slice(0, indexOfCurrentFoucs || 0),
  ].filter(
    (i) =>
      !(
        i.method === context.focus.method && i.pathId === context.focus.pathId
      ) && !alreadyHandled(i.pathId, i.method)
  );

  const next = searchOrder[0];

  return next ? { pathId: next.pathId, method: next.method } : context.focus;
}
