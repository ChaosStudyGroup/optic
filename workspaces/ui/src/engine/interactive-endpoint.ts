import { assign, spawn, Machine, StateMachine, send } from 'xstate';
import { BodyShapeDiff, ParsedDiff } from './parse-diff';
import { IShapeTrail } from './interfaces/shape-trail';
import { DiffSet } from './diff-set';
import {
  createNewRegionMachine,
  createShapeDiffMachine,
} from './interactive-diff-machine';
import { ILearnedBodies } from '@useoptic/cli-shared/build/diffs/initial-types';
import { InteractiveSessionConfig } from './interfaces/session';
import { IgnoreRule } from './interpretors/ignores/ignore-rule';

export interface InteractiveEndpointSessionStateSchema {
  states: {
    unfocused: {};
    preparing: {};
    ready: {};
  };
}

// The events that the machine handles
export type InteractiveEndpointSessionEvent =
  | {
      type: 'PREPARE';
    }
  | {
      type: 'ADD_IGNORE';
      newRule: IgnoreRule;
    }
  | {
      type: 'REMOVE_IGNORES';
      diffHash: string;
    }
  | {
      type: 'HANDLED_UPDATED';
    }
  | { type: 'RESET' };

// The context (extended state) of the machine
export interface InteractiveEndpointSessionContext {
  newRegions: {
    diffParsed: ParsedDiff;
    ref: any;
  }[];
  shapeDiffs: { diffParsed: ParsedDiff; shapeTrail: IShapeTrail; ref: any }[];
  ignored: IgnoreRule[];
  handledByDiffHash: { [key: string]: boolean };
  learningContext:
    | {
        newRegions: string;
      }
    | undefined;
}

export const newInteractiveEndpointSessionMachine = (
  pathId: string,
  method: string,
  diffs: ParsedDiff[],
  services: InteractiveSessionConfig
) => {
  return Machine<
    InteractiveEndpointSessionContext,
    InteractiveEndpointSessionStateSchema,
    InteractiveEndpointSessionEvent
  >({
    id: `${pathId}.${method}`,
    context: {
      handledByDiffHash: {},
      newRegions: [],
      shapeDiffs: [],
      ignored: [],
      learningContext: undefined,
    },
    initial: 'unfocused',
    states: {
      unfocused: {
        entry: [
          assign({
            newRegions: (context, event) => {
              const newRegionDiffs = new DiffSet(diffs, services.rfcBaseState)
                .newRegions()
                .iterator()
                .map((diffParsed, index) => {
                  const id = 'region-diff' + index;
                  return {
                    diffParsed,
                    ref: spawn(
                      createNewRegionMachine(id, diffParsed, services),
                      { name: id, autoForward: true }
                    ),
                    id,
                  };
                });
              return newRegionDiffs;
            },
            handledByDiffHash: (context) => computeHandled(context),
            shapeDiffs: (context, event) => {
              const shapeDiffsGrouped = new DiffSet(
                diffs,
                services.rfcBaseState
              ).groupedByEndpointAndShapeTrail();

              return shapeDiffsGrouped.map(
                ({ diffs, shapeDiffGroupingHash, shapeTrail }) => {
                  const id = 'shape-diff' + shapeDiffGroupingHash;
                  return {
                    shapeTrail,
                    diffParsed: diffs[0],
                    ref: spawn(createShapeDiffMachine(id, diffs[0], services), {
                      name: id,
                      autoForward: true,
                    }),
                  };
                }
              );
            },
          }),
        ],
        on: {
          PREPARE: {
            target: 'preparing',
          },
        },
      },
      preparing: {
        invoke: {
          id: 'loading-initial-regions',
          src: async (context, event) => {
            const {
              rfcService,
              rfcId,
              domainIdGenerator,
            } = services.rfcBaseState;

            let newRegionWalker: Promise<ILearnedBodies> = Promise.resolve({
              pathId,
              method,
              requests: [],
              responses: [],
            });

            if (context.newRegions.length > 0) {
              newRegionWalker = services.diffService.learnInitial(
                rfcService,
                rfcId,
                pathId,
                method,
                domainIdGenerator
              );
            }
            return await newRegionWalker;
          },
          onDone: {
            target: 'ready',
            actions: [
              assign({
                learningContext: (context, event) => {
                  return event.data;
                },
              }),
            ],
          },
        },
      },
      ready: {
        on: {
          HANDLED_UPDATED: {
            actions: assign({
              handledByDiffHash: (context) => computeHandled(context),
            }),
          },
          ADD_IGNORE: {
            actions: [
              assign({
                ignored: (context, event) => [
                  ...context.ignored,
                  event.newRule,
                ],
              }),
              (ctx) => {
                //send all ignore rules to children. they decide which ones they care about
                const notifyChildren = {
                  type: 'UPDATE_PREVIEW',
                  ignoreRules: ctx.ignored,
                };
                ctx.shapeDiffs.forEach((i) => i.ref.send(notifyChildren));
                ctx.newRegions.forEach((i) => i.ref.send(notifyChildren));
              },
            ],
          },
          RESET: {
            actions: [
              assign({
                ignored: (_, __) => [],
              }),
              (ctx) => {
                //send all ignore rules to children. they decide which ones they care about
                const notifyChildren = {
                  type: 'RESET',
                };
                ctx.shapeDiffs.forEach((i) => i.ref.send(notifyChildren));
                ctx.newRegions.forEach((i) => i.ref.send(notifyChildren));
              },
              assign({
                handledByDiffHash: (context) => computeHandled(context),
              }),
            ],
          },
          REMOVE_IGNORES: {
            actions: [
              assign({
                ignored: (context, event) =>
                  context.ignored.filter((i) => i.diffHash !== event.diffHash),
              }),
              (ctx) => {
                //send all ignore rules to children. they decide which ones they care about
                const notifyChildren = {
                  type: 'UPDATE_PREVIEW',
                  ignoreRules: ctx.ignored,
                };
                ctx.shapeDiffs.forEach((i) => i.ref.send(notifyChildren));
                ctx.newRegions.forEach((i) => i.ref.send(notifyChildren));
              },
            ],
          },
        },
      },
    },
  });
};

function computeHandled(
  context: InteractiveEndpointSessionContext
): { [key: string]: boolean } {
  const results: { [key: string]: boolean } = {};
  context.shapeDiffs.forEach((i) => {
    const value = i.ref.state.value;
    results[i.diffParsed.diffHash] = value && value.ready === 'handled';
  });
  context.newRegions.forEach((i) => {
    const value = i.ref.state.value;
    results[i.diffParsed.diffHash] = value && value.ready === 'handled';
  });
  return results;
}
