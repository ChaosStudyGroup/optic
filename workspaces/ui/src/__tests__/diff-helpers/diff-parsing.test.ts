/**
 * @jest-environment jsdom
 */

import { parseDiffsArray } from '../../engine/parse-diff';
import { DiffSet } from '../../engine/diff-set';

const diffs_raw = require('./tododiffs.json');

test('parsing a set of diffs', () => {
  const allDiffs = parseDiffsArray(diffs_raw);
  expect(allDiffs).toMatchSnapshot();
});

test('can collect new region diffs', () => {
  const diffs = new DiffSet(parseDiffsArray(diffs_raw));
  diffs
    .shapeDiffs()
    .iterator()
    .forEach((i) => {
      expect(i.raw).toMatchSnapshot();
    });
});
