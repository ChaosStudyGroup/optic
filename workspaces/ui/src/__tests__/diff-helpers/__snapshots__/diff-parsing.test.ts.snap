// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsing a set of diffs 1`] = `
Array [
  ParsedDiff {
    "diffType": "UnmatchedRequestBodyContentType",
    "interactions": Array [
      "0-2",
    ],
    "serialized_diff": Object {
      "UnmatchedRequestBodyContentType": Object {
        "interactionTrail": Object {
          "path": Array [
            Object {
              "Url": Object {},
            },
            Object {
              "Method": Object {
                "method": "POST",
              },
            },
            Object {
              "RequestBody": Object {
                "contentType": "application/json",
              },
            },
          ],
        },
        "requestsTrail": Object {
          "SpecPath": Object {
            "pathId": "path_2I5fDR2oJ4",
          },
        },
      },
    },
  },
  ParsedDiff {
    "diffType": "UnmatchedResponseBodyContentType",
    "interactions": Array [
      "0-1",
    ],
    "serialized_diff": Object {
      "UnmatchedResponseBodyContentType": Object {
        "interactionTrail": Object {
          "path": Array [
            Object {
              "Method": Object {
                "method": "PATCH",
              },
            },
            Object {
              "ResponseBody": Object {
                "contentType": "application/json",
                "statusCode": 200,
              },
            },
          ],
        },
        "requestsTrail": Object {
          "SpecPath": Object {
            "pathId": "path_ZxMkDAQ9Wf",
          },
        },
      },
    },
  },
  ParsedDiff {
    "diffType": "UnmatchedRequestBodyContentType",
    "interactions": Array [
      "0-0",
    ],
    "serialized_diff": Object {
      "UnmatchedRequestBodyContentType": Object {
        "interactionTrail": Object {
          "path": Array [
            Object {
              "Url": Object {},
            },
            Object {
              "Method": Object {
                "method": "GET",
              },
            },
          ],
        },
        "requestsTrail": Object {
          "SpecPath": Object {
            "pathId": "path_2I5fDR2oJ4",
          },
        },
      },
    },
  },
  ParsedDiff {
    "diffType": "UnmatchedResponseBodyContentType",
    "interactions": Array [
      "0-0",
    ],
    "serialized_diff": Object {
      "UnmatchedResponseBodyContentType": Object {
        "interactionTrail": Object {
          "path": Array [
            Object {
              "Method": Object {
                "method": "GET",
              },
            },
            Object {
              "ResponseBody": Object {
                "contentType": "application/json",
                "statusCode": 200,
              },
            },
          ],
        },
        "requestsTrail": Object {
          "SpecPath": Object {
            "pathId": "path_2I5fDR2oJ4",
          },
        },
      },
    },
  },
  ParsedDiff {
    "diffType": "UnmatchedRequestBodyContentType",
    "interactions": Array [
      "0-1",
    ],
    "serialized_diff": Object {
      "UnmatchedRequestBodyContentType": Object {
        "interactionTrail": Object {
          "path": Array [
            Object {
              "Url": Object {},
            },
            Object {
              "Method": Object {
                "method": "PATCH",
              },
            },
            Object {
              "RequestBody": Object {
                "contentType": "application/json",
              },
            },
          ],
        },
        "requestsTrail": Object {
          "SpecPath": Object {
            "pathId": "path_ZxMkDAQ9Wf",
          },
        },
      },
    },
  },
  ParsedDiff {
    "diffType": "UnmatchedResponseBodyContentType",
    "interactions": Array [
      "0-2",
    ],
    "serialized_diff": Object {
      "UnmatchedResponseBodyContentType": Object {
        "interactionTrail": Object {
          "path": Array [
            Object {
              "Method": Object {
                "method": "POST",
              },
            },
            Object {
              "ResponseBody": Object {
                "contentType": "application/json",
                "statusCode": 200,
              },
            },
          ],
        },
        "requestsTrail": Object {
          "SpecPath": Object {
            "pathId": "path_2I5fDR2oJ4",
          },
        },
      },
    },
  },
]
`;
